File: 2023/day19/zig/solution.zig
const std = @import("std");

const Attr = enum(u2) { x = 0, m = 1, a = 2, s = 3 };
const Op = enum(u2) { lt = 0, gt = 1, none = 2 };

const Rule = struct {
    attr: Attr,
    op: Op,
    value: i32,
    dest_idx: i32, // Index into workflow names array, or -1 for A, -2 for R
};

const Workflow = struct {
    rules: [8]Rule,
    rule_count: u8,
};

const Part = struct {
    values: [4]i32,

    fn rating(self: Part) i64 {
        var sum: i64 = 0;
        for (self.values) |v| {
            sum += v;
        }
        return sum;
    }
};

const Range = struct {
    lo: i32,
    hi: i32,

    fn count(self: Range) i64 {
        if (self.hi < self.lo) return 0;
        return @as(i64, self.hi - self.lo + 1);
    }
};

const Ranges = struct {
    r: [4]Range,

    fn init() Ranges {
        return Ranges{
            .r = [4]Range{
                .{ .lo = 1, .hi = 4000 },
                .{ .lo = 1, .hi = 4000 },
                .{ .lo = 1, .hi = 4000 },
                .{ .lo = 1, .hi = 4000 },
            },
        };
    }

    fn combinations(self: Ranges) i64 {
        var result: i64 = 1;
        for (self.r) |range| {
            result *= range.count();
        }
        return result;
    }
};

const MAX_WORKFLOWS = 600;
const MAX_PARTS = 300;

const ParsedData = struct {
    workflows: [MAX_WORKFLOWS]Workflow,
    workflow_names: [MAX_WORKFLOWS][4]u8, // Max 3 chars + padding
    workflow_count: usize,
    in_idx: i32,
    parts: [MAX_PARTS]Part,
    part_count: usize,

    fn findWorkflow(self: *const ParsedData, name: []const u8) i32 {
        if (std.mem.eql(u8, name, "A")) return -1;
        if (std.mem.eql(u8, name, "R")) return -2;

        for (0..self.workflow_count) |i| {
            const wf_name = self.workflow_names[i];
            const wf_len = for (0..4) |j| {
                if (wf_name[j] == 0) break j;
            } else 4;
            if (std.mem.eql(u8, wf_name[0..wf_len], name)) {
                return @intCast(i);
            }
        }
        return -3; // Not found
    }
};

fn parseAttr(c: u8) Attr {
    return switch (c) {
        'x' => .x,
        'm' => .m,
        'a' => .a,
        's' => .s,
        else => unreachable,
    };
}

fn parseInput(content: []const u8) ParsedData {
    var data = ParsedData{
        .workflows = undefined,
        .workflow_names = undefined,
        .workflow_count = 0,
        .in_idx = 0,
        .parts = undefined,
        .part_count = 0,
    };

    // First pass: collect all workflow names
    var sections = std.mem.splitSequence(u8, content, "\n\n");
    const workflow_section = sections.next() orelse return data;
    const parts_section = sections.next() orelse return data;

    var lines = std.mem.splitScalar(u8, workflow_section, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;
        const brace_pos = std.mem.indexOf(u8, line, "{") orelse continue;
        const name = line[0..brace_pos];

        var stored_name: [4]u8 = [4]u8{ 0, 0, 0, 0 };
        const copy_len = @min(name.len, 4);
        @memcpy(stored_name[0..copy_len], name[0..copy_len]);
        data.workflow_names[data.workflow_count] = stored_name;

        if (std.mem.eql(u8, name, "in")) {
            data.in_idx = @intCast(data.workflow_count);
        }

        data.workflow_count += 1;
    }

    // Second pass: parse rules with resolved destinations
    lines = std.mem.splitScalar(u8, workflow_section, '\n');
    var wf_idx: usize = 0;
    while (lines.next()) |line| {
        if (line.len == 0) continue;
        const brace_pos = std.mem.indexOf(u8, line, "{") orelse continue;

        const rules_str = line[brace_pos + 1 .. line.len - 1];
        var workflow = Workflow{ .rules = undefined, .rule_count = 0 };

        var rule_iter = std.mem.splitScalar(u8, rules_str, ',');
        while (rule_iter.next()) |rule_str| {
            if (std.mem.indexOf(u8, rule_str, ":")) |colon_pos| {
                const condition = rule_str[0..colon_pos];
                const destination = rule_str[colon_pos + 1 ..];

                workflow.rules[workflow.rule_count] = Rule{
                    .attr = parseAttr(condition[0]),
                    .op = if (condition[1] == '<') .lt else .gt,
                    .value = std.fmt.parseInt(i32, condition[2..], 10) catch 0,
                    .dest_idx = data.findWorkflow(destination),
                };
            } else {
                workflow.rules[workflow.rule_count] = Rule{
                    .attr = .x,
                    .op = .none,
                    .value = 0,
                    .dest_idx = data.findWorkflow(rule_str),
                };
            }
            workflow.rule_count += 1;
        }

        data.workflows[wf_idx] = workflow;
        wf_idx += 1;
    }

    // Parse parts
    var part_lines = std.mem.splitScalar(u8, parts_section, '\n');
    while (part_lines.next()) |line| {
        if (line.len == 0 or line[0] != '{') continue;

        var part = Part{ .values = [4]i32{ 0, 0, 0, 0 } };
        const inner = line[1 .. line.len - 1];

        var attr_iter = std.mem.splitScalar(u8, inner, ',');
        while (attr_iter.next()) |attr_str| {
            const attr = parseAttr(attr_str[0]);
            const value = std.fmt.parseInt(i32, attr_str[2..], 10) catch 0;
            part.values[@intFromEnum(attr)] = value;
        }

        data.parts[data.part_count] = part;
        data.part_count += 1;
    }

    return data;
}

fn processPart(data: *const ParsedData, part: Part) bool {
    var current: i32 = data.in_idx;

    while (current >= 0) {
        const workflow = data.workflows[@intCast(current)];

        for (workflow.rules[0..workflow.rule_count]) |rule| {
            if (rule.op == .none) {
                current = rule.dest_idx;
                break;
            } else {
                const part_val = part.values[@intFromEnum(rule.attr)];
                const matches = switch (rule.op) {
                    .lt => part_val < rule.value,
                    .gt => part_val > rule.value,
                    .none => unreachable,
                };
                if (matches) {
                    current = rule.dest_idx;
                    break;
                }
            }
        }
    }

    return current == -1; // -1 = A (accepted)
}

fn part1(data: *const ParsedData) i64 {
    var total: i64 = 0;
    for (data.parts[0..data.part_count]) |part| {
        if (processPart(data, part)) {
            total += part.rating();
        }
    }
    return total;
}

fn countAccepted(data: *const ParsedData, workflow_idx: i32, ranges: Ranges) i64 {
    if (workflow_idx == -2) return 0; // R (rejected)
    if (workflow_idx == -1) return ranges.combinations(); // A (accepted)

    const workflow = data.workflows[@intCast(workflow_idx)];

    var total: i64 = 0;
    var current_ranges = ranges;

    for (workflow.rules[0..workflow.rule_count]) |rule| {
        if (rule.op == .none) {
            total += countAccepted(data, rule.dest_idx, current_ranges);
        } else {
            const attr_idx = @intFromEnum(rule.attr);
            const lo = current_ranges.r[attr_idx].lo;
            const hi = current_ranges.r[attr_idx].hi;

            if (rule.op == .lt) {
                if (lo < rule.value) {
                    var new_ranges = current_ranges;
                    new_ranges.r[attr_idx] = .{ .lo = lo, .hi = @min(hi, rule.value - 1) };
                    total += countAccepted(data, rule.dest_idx, new_ranges);
                }
                if (hi >= rule.value) {
                    current_ranges.r[attr_idx] = .{ .lo = @max(lo, rule.value), .hi = hi };
                } else {
                    break;
                }
            } else {
                if (hi > rule.value) {
                    var new_ranges = current_ranges;
                    new_ranges.r[attr_idx] = .{ .lo = @max(lo, rule.value + 1), .hi = hi };
                    total += countAccepted(data, rule.dest_idx, new_ranges);
                }
                if (lo <= rule.value) {
                    current_ranges.r[attr_idx] = .{ .lo = lo, .hi = @min(hi, rule.value) };
                } else {
                    break;
                }
            }
        }
    }

    return total;
}

fn part2(data: *const ParsedData) i64 {
    return countAccepted(data, data.in_idx, Ranges.init());
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    const data = parseInput(content);

    std.debug.print("Part 1: {d}\n", .{part1(&data)});
    std.debug.print("Part 2: {d}\n", .{part2(&data)});
}

File: 2023/day19/rust/solution.rs
// Day 19: Aplenty - Workflow processing and range analysis

use std::collections::HashMap;
use std::fs;

#[derive(Clone, Debug)]
enum Rule {
    Conditional {
        attr: char,
        op: char,
        value: i64,
        destination: String,
    },
    Default {
        destination: String,
    },
}

type Workflows = HashMap<String, Vec<Rule>>;

#[derive(Clone, Debug)]
struct Part {
    x: i64,
    m: i64,
    a: i64,
    s: i64,
}

impl Part {
    fn get(&self, attr: char) -> i64 {
        match attr {
            'x' => self.x,
            'm' => self.m,
            'a' => self.a,
            's' => self.s,
            _ => panic!("Invalid attribute: {}", attr),
        }
    }

    fn sum(&self) -> i64 {
        self.x + self.m + self.a + self.s
    }
}

#[derive(Clone, Debug)]
struct Ranges {
    x: (i64, i64),
    m: (i64, i64),
    a: (i64, i64),
    s: (i64, i64),
}

impl Ranges {
    fn new() -> Self {
        Ranges {
            x: (1, 4000),
            m: (1, 4000),
            a: (1, 4000),
            s: (1, 4000),
        }
    }

    fn get(&self, attr: char) -> (i64, i64) {
        match attr {
            'x' => self.x,
            'm' => self.m,
            'a' => self.a,
            's' => self.s,
            _ => panic!("Invalid attribute: {}", attr),
        }
    }

    fn set(&mut self, attr: char, range: (i64, i64)) {
        match attr {
            'x' => self.x = range,
            'm' => self.m = range,
            'a' => self.a = range,
            's' => self.s = range,
            _ => panic!("Invalid attribute: {}", attr),
        }
    }

    fn combinations(&self) -> i64 {
        let count = |r: (i64, i64)| -> i64 { (r.1 - r.0 + 1).max(0) };
        count(self.x) * count(self.m) * count(self.a) * count(self.s)
    }
}

fn parse_input(content: &str) -> (Workflows, Vec<Part>) {
    let sections: Vec<&str> = content.trim().split("\n\n").collect();
    let workflow_section = sections[0];
    let parts_section = sections[1];

    // Parse workflows
    let mut workflows = HashMap::new();
    for line in workflow_section.lines() {
        let brace_pos = line.find('{').unwrap();
        let name = line[..brace_pos].to_string();
        let rules_str = &line[brace_pos + 1..line.len() - 1]; // Remove { and }

        let mut rules = Vec::new();
        for rule_str in rules_str.split(',') {
            if let Some(colon_pos) = rule_str.find(':') {
                let condition = &rule_str[..colon_pos];
                let destination = rule_str[colon_pos + 1..].to_string();

                let attr = condition.chars().next().unwrap();
                let op = condition.chars().nth(1).unwrap();
                let value: i64 = condition[2..].parse().unwrap();

                rules.push(Rule::Conditional {
                    attr,
                    op,
                    value,
                    destination,
                });
            } else {
                rules.push(Rule::Default {
                    destination: rule_str.to_string(),
                });
            }
        }
        workflows.insert(name, rules);
    }

    // Parse parts
    let mut parts = Vec::new();
    for line in parts_section.lines() {
        let inner = &line[1..line.len() - 1]; // Remove { and }
        let mut part = Part {
            x: 0,
            m: 0,
            a: 0,
            s: 0,
        };

        for assignment in inner.split(',') {
            let mut parts_iter = assignment.split('=');
            let attr = parts_iter.next().unwrap();
            let value: i64 = parts_iter.next().unwrap().parse().unwrap();
            match attr {
                "x" => part.x = value,
                "m" => part.m = value,
                "a" => part.a = value,
                "s" => part.s = value,
                _ => panic!("Unknown attribute: {}", attr),
            }
        }
        parts.push(part);
    }

    (workflows, parts)
}

fn process_part(workflows: &Workflows, part: &Part) -> bool {
    let mut current = "in".to_string();

    while current != "A" && current != "R" {
        let rules = workflows.get(&current).unwrap();

        for rule in rules {
            match rule {
                Rule::Default { destination } => {
                    current = destination.clone();
                    break;
                }
                Rule::Conditional {
                    attr,
                    op,
                    value,
                    destination,
                } => {
                    let part_value = part.get(*attr);
                    let matches = match op {
                        '<' => part_value < *value,
                        '>' => part_value > *value,
                        _ => panic!("Unknown operator: {}", op),
                    };
                    if matches {
                        current = destination.clone();
                        break;
                    }
                }
            }
        }
    }

    current == "A"
}

fn part1(workflows: &Workflows, parts: &[Part]) -> i64 {
    parts
        .iter()
        .filter(|p| process_part(workflows, p))
        .map(|p| p.sum())
        .sum()
}

fn count_accepted(workflows: &Workflows, workflow: &str, ranges: Ranges) -> i64 {
    if workflow == "R" {
        return 0;
    }
    if workflow == "A" {
        return ranges.combinations();
    }

    let mut total = 0;
    let mut ranges = ranges;
    let rules = workflows.get(workflow).unwrap();

    for rule in rules {
        match rule {
            Rule::Default { destination } => {
                total += count_accepted(workflows, destination, ranges.clone());
            }
            Rule::Conditional {
                attr,
                op,
                value,
                destination,
            } => {
                let (lo, hi) = ranges.get(*attr);

                match op {
                    '<' => {
                        // Split: [lo, value-1] goes to destination, [value, hi] continues
                        if lo < *value {
                            let mut new_ranges = ranges.clone();
                            new_ranges.set(*attr, (lo, hi.min(*value - 1)));
                            total += count_accepted(workflows, destination, new_ranges);
                        }
                        if hi >= *value {
                            ranges.set(*attr, (lo.max(*value), hi));
                        } else {
                            break;
                        }
                    }
                    '>' => {
                        // Split: [value+1, hi] goes to destination, [lo, value] continues
                        if hi > *value {
                            let mut new_ranges = ranges.clone();
                            new_ranges.set(*attr, (lo.max(*value + 1), hi));
                            total += count_accepted(workflows, destination, new_ranges);
                        }
                        if lo <= *value {
                            ranges.set(*attr, (lo, hi.min(*value)));
                        } else {
                            break;
                        }
                    }
                    _ => panic!("Unknown operator: {}", op),
                }
            }
        }
    }

    total
}

fn part2(workflows: &Workflows) -> i64 {
    count_accepted(workflows, "in", Ranges::new())
}

fn main() {
    let content = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let (workflows, parts) = parse_input(&content);

    println!("Part 1: {}", part1(&workflows, &parts));
    println!("Part 2: {}", part2(&workflows));
}
